src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTryFinally.cs

```diff
@@ -3758,5 +3759,492 @@ .maxstack  1
 }
 ");
         }
+
+        [Fact]
+        [WorkItem("https://github.com/dotnet/roslyn/issues/67494")]
+        public void WhenWithAlwaysThrowingExpression_01()
+        {
+            var source = """
+class C
+{
+    static bool _throw;
+
+    static void Main()
+    {
+        M();
+
+        _throw = true;
+
+        try 
+        {
+            M();
+        }
+        catch (System.NullReferenceException)
+        {
+            System.Console.Write("Catch");
+        }
+    }
+
+    static void M()
+    {
+        try
+        {
+            M1();
+        }
+        catch when (true ? throw M2() : true)
+        {
+            M3();
+        }
+
+        M4();
+    }
+
+    static void M1()
+    {
+        System.Console.Write("M1");
+        if (_throw) throw null;
+    }
+
+    static System.Exception M2()
+    {
+        System.Console.Write("M2");
+        return new System.NotSupportedException();
+    }
+
+    static void M3()
+    {
+        System.Console.Write("M2");
+    }
+
+    static void M4()
+    {
+        System.Console.Write("M4");
+    }
+}
+""";
+
+            var verifier = CompileAndVerify(source, options: TestOptions.ReleaseExe, expectedOutput: "M1M4M1M2Catch",
+                // False PEVerify failure:
+                // PEVerify failed for assembly
+                // [ : C::M][mdToken=0x6000004][offset 0x0000000E] Stack not empty when leaving an exception filter.
```

Comment:
I think the PEVerify failures can be verified programmatically by something like `Verification.FailsPEVerify with { PEVerifyMessage = "..." }`.